### ISSUE TITLE
catalog: limit entity read access to certain users/groups

### ISSUE THREAD

**Issue by freben** [backstage/backstage#3218]
## Feature Suggestion

Internally at Spotify we have had a very open mindset, where everybody can see pretty much all information available in Backstage, at least for reading - while management/write operations are limited to those that have the appropriate rights.

This is not good enough in some environments. There are many possible reasons for wanting to limit the access to information to certain groups of people, and the catalog is one such place where it would be possible to glean sensitive information by spidering through all of the entities that are stored inside.

There should be a mechanism for users to limit access to their entities.

## Scope and Implications

The current assumption is that entities should be treated as completely absent, for users that do not have the correct permission level, such that there's no way to even get a hint of their existence through for example name-guessing. Just the name itself may reveal sensitive information, for example if it hints at a not-yet-disclosed upcoming acquisition or research product of a sensitive nature.

This does have implications on several levels, making the implementation nontrivial.

- The basic (and possibly easiest to achieve) need is that when listing entities or attempting to fetch a single entity, the user should see a behaviour exactly as if the entity does not exist, if they do not have permissions to see it.
- When the user tries to register a new entity, that collides with a secret (hidden) entity, the system will have to communicate that in a sane way to the caller, that does not reveal too much information. For example if the message states "you do not have permission to overwrite the API competitor-api" or "you do not have access to accidentally-acquisition-colliding-namespace" or similar, it does leak information.
- When there are bidirectional references between entities A and B, and the user does not have permissions to see B, then they should not see those relations while fetching A.
- The above presents a problem if the relation came from the `spec` - for example, saying `implementsApis : [othernamespace/secretthing]` since we likely cannot reliably censor the spec data. But this must be regarded as a misconfiguration; this would mean that the user placed sensitive data in a public place, so we probably won't handle that corner case.
- This may also extend to registered catalog locations, not just entities. The act of registering a sensitive thing from a secret location may be considered sensitive as well.
- There may be other information sinks that have to be handled one by one. For example, if there is an event bus originating from the catalog that publishes events as entities are registered, changed and deleted, then those events will potentially contain sensitive information as well.

## Authorization (out of scope)

This all overlaps with the concept of authorization. If you do not have permissions to view an entity, you likely should not be permitted to delete or modify it either. This is intentionally left out of scope for this discussion, however, the design of these concepts should most likely also take future authorization needs into account so that the concepts can be unified if deemed appropriate.

## Possible Implementation

We are still not assigning much logic to the entity `metadata.namespace` concept. A likely useful abstraction level could be to allow the binding of ACLs, RBAC, or ABAC onto the namespace level. This, coupled with leaving the `default` namespace open for all per default, would make it a fairly clear user action to make a claim that "users of teams `team1` and `team2` have read rights on namespace `ns3`".

## Alternative Implementations

- Security conscious users could install several distinct catalogs, each with separate auth requirements. There could be a low security base catalog A, that contains all of the insensitive data. Then a walled off catalog B could house the sensitive data, plus have a processor in place that also ingests all of the data from A, making B a superset of A without A ever noticing. This would make the implementation trivial (a simple catalog-reader-processor class needs to be written), and gives stronger guarantees and better isolation, at the cost of possibly higher maintenance cost. Also, it works well for the write case, but what if a user that has access to B wants to register a new location - they must then be given the choice of whether they want to do it in catalog A or B. This applies to all write related operations. This places a greater burden on the frontend implementation.

## Context

See also discussion in #1401.

**Comment by OrkoHunter** [comment]
Re-opening since I think this isn't really stale and ties into RBAC implementations.

**Comment by freben** [comment]
@timbonicus I think that we could consider this closable after #7761 landed. Agree?

**Comment by timbonicus** [comment]
@freben There are some good points in the initial description that should be vetted against the outcome of #7761. Closing this, but I made a note to revisit these points during the catalog authorization integration.
